spanconfig: introduce spanconfig.Splitter

irfansharif opened the pull request:
Splitter returns the set of all possible split points for the given
table descriptor. It steps through every "unit" that we can apply
configurations over (table, indexes, partitions and sub-partitions) and
figures out the actual key boundaries that we may need to split over.
For example:
```
  CREATE TABLE db.parts(i INT PRIMARY KEY, j INT) PARTITION BY LIST (i) (
    PARTITION one_and_five    VALUES IN (1, 5),
    PARTITION four_and_three  VALUES IN (4, 3),
    PARTITION everything_else VALUES IN (6, default)
  );
```
Assuming a table ID of 108, we'd generate:
```
  /Table/108
  /Table/108/1
  /Table/108/1/1
  /Table/108/1/2
  /Table/108/1/3
  /Table/108/1/4
  /Table/108/1/5
  /Table/108/1/6
  /Table/108/1/7
  /Table/108/2
```

This is going to serve as the underlying library for #70555. In a future
commit we'll introduce two system tables, one for all tenants (both
guest and host) and one on just the host.

The first will be maintained by the SQL pod, using the library, as part
of every schema change operation. We'll record in it a single row
containing the total number of splits the pod's currently using up by
comparing the before/after of a descriptor operation to see if the
commit would add or reduce the number of necessary splits. We'll abort
the txn if it's over a pre-set limit (INF for the host tenant, and using
a KV-settable-only cluster setting for guests: #73857).

The second table will be used only for KV's internal book keeping,
sitting behind the KVAccessor interface. We'll use it to hard reject
RPCs installing span configs that induce splits more than a tenant is
allowed to. With the SQL pods coordinating using the KV-set limits, this
would only happen as a result of faulty SQL code.

Put together, we'll have a co-operative form of per-tenant split limits.

Release note: None

irfansharif committed:
spanconfig: introduce spanconfig.Splitter

Splitter returns the set of all possible split points for the given
table descriptor. It steps through every "unit" that we can apply
configurations over (table, indexes, partitions and sub-partitions) and
figures out the actual key boundaries that we may need to split over.
For example:

  CREATE TABLE db.parts(i INT PRIMARY KEY, j INT) PARTITION BY LIST (i) (
    PARTITION one_and_five    VALUES IN (1, 5),
    PARTITION four_and_three  VALUES IN (4, 3),
    PARTITION everything_else VALUES IN (6, default)
  );

Assuming a table ID of 108, we'd generate:

  /Table/108
  /Table/108/1
  /Table/108/1/1
  /Table/108/1/2
  /Table/108/1/3
  /Table/108/1/4
  /Table/108/1/5
  /Table/108/1/6
  /Table/108/1/7
  /Table/108/2

---

This is going to serve as the underlying library for #70555. In a future
commit we'll introduce two system tables, one for all tenants (both
guest and host) and one on just the host.

The first will be maintained by the SQL pod, using the library, as part
of every schema change operation. We'll record in it a single row
containing the total number of splits the pod's currently using up by
comparing the before/after of a descriptor operation to see if the
commit would add or reduce the number of necessary splits. We'll abort
the txn if it's over a pre-set limit (INF for the host tenant, and using
a KV-settable-only cluster setting for guests: #73857).

The second table will be used only for KV's internal book keeping,
sitting behind the KVAccessor interface. We'll use it to hard reject
RPCs installing span configs that induce splits more than a tenant is
allowed to. With the SQL pods coordinating using the KV-set limits, this
would only happen as a result of faulty SQL code.

Put together, we'll have a co-operative form of per-tenant split limits.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75803)
<!-- Reviewable:end -->

irfansharif committed:
Makefile: adopt a TODO

The 'XXX:' is a useful tag to configure IDEs/editors to track changes
you want to address pre-push. This commit adopts a checked-in instance
of the tag for this author's convenience.

Release note: None

irfansharif commented:
Tried capturing a few interest partitioning variants, but super open to others. Our syntax makes for far too many degrees of flexibility and it's likely I've missed edge cases (or at least testing them explicitly).

arulajmani reviewed (commented):
This is quite cool! I took a first pass and this mostly looks good. I do mean to come back and examine some of the stuff around `splitInner` more closely. 

> Reviewed 1 of 1 files at r1, 11 of 13 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-:-MvvE_-13TGT2NKw0bmp:b-15xnch)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, and @nvanbenschoten)

___
*[Makefile, line 1374 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-MvuuiNj7kVzTKUj5vCF:-MvuuiNj7kVzTKUj5vCG:b-bpotc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f985db0a3370395267715f155183fb2075d621e/Makefile#L1374)):*
> ```Makefile
> # FOO-manifest.js were both out-of-date. [0]
> #
> # TODO(irfansharif): Ideally we'd scope the dependency on $(UI_PROTOS*) to the
> ```

Maybe "NB" this instead of adopting the TODO, unless you're meaning to actually do something about this in the future?
___
*[pkg/spanconfig/spanconfig.go, line 300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvv10AQALtuTszoXNoj:-Mvv10AQALtuTszoXNok:b-ba96ru) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfig.go#L300)):*
> ```Go
> //   span, we're missing /Table/106{-/1}, /Table/106/{2-3}, /Table/10{6/4-7}.
> //
> // For N children:
> ```

This is a really nice observation! Would this approach be preferable over the current implementation? 
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 66 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-MvuxUdu9xOCnuIicOVw:-MvuxUdu9xOCnuIicOVx:b-v7kpeo) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L66)):*
> ```Go
> 	var tableSpans []roachpb.Span
> 	var lastIndexSpan roachpb.Span
> 	if err := catalog.ForEachIndex(table, catalog.IndexOpts{}, func(index catalog.Index) error {
> ```

Should this be `ForEachNonDropIndex` instead?
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvuxh4O8ICeLE9WoI0H:-Mvuxh4O8ICeLE9WoI0I:b-y9h9m8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L68)):*
> ```Go
> 	if err := catalog.ForEachIndex(table, catalog.IndexOpts{}, func(index catalog.Index) error {
> 		currentIndexSpan := table.IndexSpan(s.codec, index.GetID())
> 		if index.Ordinal() == 0 {
> ```

nit: index.Primary() instead?
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 95 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvuy7BB3iYhIsxVMuGL:-Mvuy7BB3iYhIsxVMuGM:bxxj24y) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L95)):*
> ```Go
> 		}
> 
> 		// TODO(irfansharif): Should we only assert under test builds?
> ```

Could use `ReportOrPanic` here. Also, consider pulling this validation logic into a function. 
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 133 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvv-gPx4Lgtb_U2TstZ:-Mvv-gPx4Lgtb_U2Tst_:bufy3o3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L133)):*
> ```Go
> }
> 
> func (s *Splitter) splitInner(ctx context.Context, part partition) ([]roachpb.Span, error) {
> ```

I haven't looked at this very closely, and plan to come back to it, but at first glance this logic seems similar to what we do to generate subzone spans (and `indexCoveringsForPartitioning`). Did you consider re-using that codepath or unifying the two together, instead of this parallel implementation?
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/indexes, line 12 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-MvvDll-1-fVlpP9CHhs:-MvvDll-1-fVlpP9CHht:b-a0s139) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/indexes#L12)):*
> ```
> /Table/106/1
> /Table/106/2
> /Table/106/3
> ```

Should we test dropping a secondary index as well, so that the index's ID leaves a "hole" behind? 
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-MvvDT_LAhdyRxqZRYRb:-MvvDT_LAhdyRxqZRYRc:b6lqhjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions#L1)):*
> ```
> exec-sql
> ```

You could also test some partitioning variants on secondary indexes. It might also be worth adding a test for implicit partitioning as well. 
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions, line 176 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-MvvCiHI0xcRgOSI81b_:-MvvCiHI0xcRgOSI81ba:bici3dr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions#L176)):*
> ```
> 
> 
> exec-sql
> ```

Consider adding some words about each of these partitioning constructions that we're testing over here -- it might be useful if we've to ever come back to these. 
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/datadriven_test.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvv1U4hCaZ7VYwdNoFU:-Mvv1U4hCaZ7VYwdNoFV:bk8dl0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/ccl/spanconfigccl/spanconfigsplitterccl/datadriven_test.go#L35)):*
> ```Go
> )
> 
> // TestDataDriven is a data-driven test for the spanconfig.SQLTranslator. It
> ```

spanconfig.Splitter? Speaking of the SQLTranslator, have we compared its output to that of the Splitter? Are the number of splits/split points the same if we set zone configurations on all indexes/partitions/subpartitions? Thoughts on testing something like this?
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/datadriven_test.go, line 121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvv3Oy2B0MlHRFlpEd8:-Mvv3Oy2B0MlHRFlpEd9:b-6xc0px) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/ccl/spanconfigccl/spanconfigsplitterccl/datadriven_test.go#L121)):*
> ```Go
> 					objID = tenant.LookupTableByName(ctx, dbName, tbName).GetID()
> 				default:
> 					d.Fatalf(t, "insufficient/improper args (%v) provided to translate", d.CmdArgs)
> ```

s/translate/splits


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
I like this very much!

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-:-Mvyw7Gk2gDffoKKUwJG:b-n53h6p)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @irfansharif, and @nvanbenschoten)

___
*[pkg/spanconfig/spanconfig.go, line 300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvv10AQALtuTszoXNoj:-MvysrT47KySeo7hQlnq:bwkl7sp) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfig.go#L300)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This is a really nice observation! Would this approach be preferable over the current implementation? 
</blockquote></details>

I imagine the current approach makes it easier to test and debug. I'd support only exposing the count at this layer and having an internal API somewhere that does render the keys too and ensures that that matches the count. 
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-MvytCTV4p7JNhoLZvKv:-MvytCTV4p7JNhoLZvKw:b-3tel7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L33)):*
> ```Go
> // Splitter is the concrete implementation of spanconfig.Splitter.
> type Splitter struct {
> 	execCfg *sql.ExecutorConfig
> ```

the `execCfg` is a hefty dependency and usually is something I think of as a code smell, especially if you hold on to the whole beast. Reading the code, I don't see you use it literally anywhere. 
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 34 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvyvp_A0VnGb08CzKan:-Mvyvp_A0VnGb08CzKao:b-8cbbxp) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L34)):*
> ```Go
> type Splitter struct {
> 	execCfg *sql.ExecutorConfig
> 	alloc   tree.DatumAlloc
> ```

do we like that this is not safe for concurrent use? My 2c is that we should just put the DatumAlloc on the stack 
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvyth3_1-UbbSgf-KP0:-Mvyth3_1-UbbSgf-KP1:bjj664i) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions#L1)):*
> ```
> exec-sql
> ```

can you test `REGIONAL BY ROW` tables and `PARTITION ALL BY` implicit partitioning? 


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
Added a bit more commentary and more test cases, PTAL. Do we think this is ready to merge?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-:-Mw4jpG4-ueuPH_GLBlN:b-odtlax)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @nvanbenschoten)

___
*[Makefile, line 1374 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-MvuuiNj7kVzTKUj5vCF:-Mvxk8zT3C3_EWiEYM5L:b-fi5bg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f985db0a3370395267715f155183fb2075d621e/Makefile#L1374)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Maybe "NB" this instead of adopting the TODO, unless you're meaning to actually do something about this in the future?
</blockquote></details>

Planning to delete the entire Makefile :stuck_out_tongue: 
___
*[pkg/spanconfig/spanconfig.go, line 300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvv10AQALtuTszoXNoj:-MvxkJcPDCGuUFL8q4W6:b-e08w5w) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfig.go#L300)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I imagine the current approach makes it easier to test and debug. I'd support only exposing the count at this layer and having an internal API somewhere that does render the keys too and ensures that that matches the count. 
</blockquote></details>

Yea, I think testing the internals is less clearer (can't actually print out the split keys in the way we do now) and it's not more/less efficient, I don't think, since we'd still have to recurse all the way through. I did write this version up first but then to write the tests, I was writing comments explaining the split count, so I just did that instead.

 We could skip decoding tuples but I doubt it'll be a bottle neck since they're just byte slice ops. Also for PARTITION BY LISTs and indexes where index IDs _are_ adjacent (they'd classify as "non-contiguous" in the general case, but could be contiguous depending on the actual values) we'd be returning a higher split count than necessary. I don't think that really matters as we're already assuming the worst with all partitions/indexes/base table having unique configs, but it's a consideration. 

Anyway, I think it's fine as is.
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-MvytCTV4p7JNhoLZvKv:-Mw1zjLX7or19aA0WQMo:b-rqeyv) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L33)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

the `execCfg` is a hefty dependency and usually is something I think of as a code smell, especially if you hold on to the whole beast. Reading the code, I don't see you use it literally anywhere. 
</blockquote></details>

Woops, just cargo culting. Removed.
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 34 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvyvp_A0VnGb08CzKan:-Mw21Lwy-jdBtVWvcJU0:bnryw31) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L34)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

do we like that this is not safe for concurrent use? My 2c is that we should just put the DatumAlloc on the stack 
</blockquote></details>

Done.

```go
  alloc := &tree.DatumAlloc{}
  _, fromKey, err := rowenc.DecodePartitionTuple(
    alloc, s.codec, part.table, part.index, part, from, part.prefixDatums)
  if err != nil {
    return err
  }
```

```
$ go build -gcflags '-m' github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigsplitter 
[...]
../../../../spanconfig/spanconfigsplitter/splitter.go:141:13: &tree.DatumAlloc{} does not escape
```
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 66 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-MvuxUdu9xOCnuIicOVw:-Mw24W8wBNO5nrCmNn4F:b-98c33k) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L66)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should this be `ForEachNonDropIndex` instead?
</blockquote></details>

I don't know? Is there a test you have in mind that would break if we didn't use ForEachIndex? Looking at GenerateSubzoneSpans, the other place where we work through partition spans, we also use ForEachIndex. That said, we also specify `AddMutations: true`. Do either of you (+cc @ajwerner) know what that means/why?

https://github.com/cockroachdb/cockroach/blob/2b7ba8c72bd5031a5ae28945b5906ef0d407e6be/pkg/sql/partition_utils.go#L112-L114
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvuxh4O8ICeLE9WoI0H:-MvxlX528gLwJNg6uhD6:bjrytc0) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L68)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: index.Primary() instead?
</blockquote></details>

ForEachIndex iterates in Ordinal order, and .Primary() is the same as .Ordinal() == 0. I like this form cause it parallels the partition iteration below where we also do special gap constructions depending on whether we're the first "segment" or not.
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 95 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvuy7BB3iYhIsxVMuGL:-Mw25wt8AMOEwmE8tPYV:bw559o7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L95)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Could use `ReportOrPanic` here. Also, consider pulling this validation logic into a function. 
</blockquote></details>

From `ReportOrPanic`:

```
// binary, or panics, if triggered in tests. This is intended to be used for
// failing assertions which are recoverable but serious enough to report and to
// cause tests to fail.
```

I always have a hard time knowing what to do with these kinds of APIs honestly. Would these assertions be "recoverable"? Something is seriously buggy if they get tripped, and improper accounting based on this library may get you into a very hairy mess where you've committed schemas that KV will reject span configs over. I don't know if that's better. I hope we don't trip these assertions, but if we do, I don't know if it's good to let the process proceed.
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 133 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvv-gPx4Lgtb_U2TstZ:-Mw26acPApQCw1AuS72x:b96jeq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L133)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I haven't looked at this very closely, and plan to come back to it, but at first glance this logic seems similar to what we do to generate subzone spans (and `indexCoveringsForPartitioning`). Did you consider re-using that codepath or unifying the two together, instead of this parallel implementation?
</blockquote></details>

I still don't really understand that code, and I think this green-room implementation is simpler/cleaner. Re-using those fragments I think will only add to the confusion and feels like false economy. The next time anyone has to deal with partitioning descriptors, I hope they look at this parallel implementation to figure out what the hell is going on.
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/indexes, line 12 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-MvvDll-1-fVlpP9CHhs:-Mw271OA-dxZp1bDfpl6:b-fznogt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/indexes#L12)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we test dropping a secondary index as well, so that the index's ID leaves a "hole" behind? 
</blockquote></details>

Good idea, done.
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-MvvDT_LAhdyRxqZRYRb:-Mw4fdkS3Lveu8tg-4-N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions#L1)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

You could also test some partitioning variants on secondary indexes. It might also be worth adding a test for implicit partitioning as well. 
</blockquote></details>

Done.
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvyth3_1-UbbSgf-KP0:-Mw4iQKG9kdgHFvOLCD-:b-nv47pz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions#L1)):*
> PARTITION ALL BY implicit partitioning?

Done.

Testing REGIONAL BY ROW tables will complicate the test setup here -- will need to set up a MR cluster with the right regions. Once we integrate this into the schema changer, it'll be easier to test with our existing logic tests that do decent MR setup already. If it's all the same, I'll come back to it then.
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions, line 176 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-MvvCiHI0xcRgOSI81b_:-Mw4ePaK3pPU5-HCdvLY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions#L176)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Consider adding some words about each of these partitioning constructions that we're testing over here -- it might be useful if we've to ever come back to these. 
</blockquote></details>

Done.
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/datadriven_test.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvv1U4hCaZ7VYwdNoFU:-Mvy0sUG0Dmj5ZZoD1CF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/ccl/spanconfigccl/spanconfigsplitterccl/datadriven_test.go#L35)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

spanconfig.Splitter? Speaking of the SQLTranslator, have we compared its output to that of the Splitter? Are the number of splits/split points the same if we set zone configurations on all indexes/partitions/subpartitions? Thoughts on testing something like this?
</blockquote></details>

Done.
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/datadriven_test.go, line 121 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvv3Oy2B0MlHRFlpEd8:-Mw274B80gBb5ZTvq2g2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/ccl/spanconfigccl/spanconfigsplitterccl/datadriven_test.go#L121)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

s/translate/splits
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
Thanks for adding the commentary around test cases. Only a few minor comments from me, should be ready to merge once you address them. 

> Reviewed 13 of 13 files at r3, 13 of 15 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-:-Mw7MKXo29wvQ7mNOGjx:buo9yby)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @irfansharif, and @nvanbenschoten)

___
*[pkg/spanconfig/spanconfig.go, line 300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvv10AQALtuTszoXNoj:-Mw7Agqv9N2LpZQFg_Tq:b92m471) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfig.go#L300)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Yea, I think testing the internals is less clearer (can't actually print out the split keys in the way we do now) and it's not more/less efficient, I don't think, since we'd still have to recurse all the way through. I did write this version up first but then to write the tests, I was writing comments explaining the split count, so I just did that instead.

 We could skip decoding tuples but I doubt it'll be a bottle neck since they're just byte slice ops. Also for PARTITION BY LISTs and indexes where index IDs _are_ adjacent (they'd classify as "non-contiguous" in the general case, but could be contiguous depending on the actual values) we'd be returning a higher split count than necessary. I don't think that really matters as we're already assuming the worst with all partitions/indexes/base table having unique configs, but it's a consideration. 

Anyway, I think it's fine as is.
</blockquote></details>

Let's change the interface here to return only a count and move the current implementation and testing a layer below it. All usages of this thing are going to be of the form `len(Splitter.Splits())` (they are, right?) so that feels cleaner. 

Also, could you please move the TODO around the alternative implementation to package `spanconfigsplitter` now that we're not going with that approach and it's an implementation detail? 
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 66 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-MvuxUdu9xOCnuIicOVw:-Mw7CHAp7P8sRQHYCjxP:b-urve8m) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L66)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

I don't know? Is there a test you have in mind that would break if we didn't use ForEachIndex? Looking at GenerateSubzoneSpans, the other place where we work through partition spans, we also use ForEachIndex. That said, we also specify `AddMutations: true`. Do either of you (+cc @ajwerner) know what that means/why?

https://github.com/cockroachdb/cockroach/blob/2b7ba8c72bd5031a5ae28945b5906ef0d407e6be/pkg/sql/partition_utils.go#L112-L114
</blockquote></details>

I was thinking we might be over counting if we count splits for indexes that are being dropped. Maybe that's okay, considering the drop can fail, and you don't want to run into limits when that happens. 
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 95 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvuy7BB3iYhIsxVMuGL:-Mw7Lsbw74VOKKPWHqQD:byyafl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L95)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

From `ReportOrPanic`:

```
// binary, or panics, if triggered in tests. This is intended to be used for
// failing assertions which are recoverable but serious enough to report and to
// cause tests to fail.
```

I always have a hard time knowing what to do with these kinds of APIs honestly. Would these assertions be "recoverable"? Something is seriously buggy if they get tripped, and improper accounting based on this library may get you into a very hairy mess where you've committed schemas that KV will reject span configs over. I don't know if that's better. I hope we don't trip these assertions, but if we do, I don't know if it's good to let the process proceed.
</blockquote></details>

On the flip side, you could imagine being well under your limits and this getting tripped up, in which case you don't want to block the ability to run schema changes because of it. I'm fine with us being ultra defensive and fataling here too. 
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 133 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvv-gPx4Lgtb_U2TstZ:-Mw7IxMtA1vuYreqrkAq:b-e3e2fi) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L133)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

I still don't really understand that code, and I think this green-room implementation is simpler/cleaner. Re-using those fragments I think will only add to the confusion and feels like false economy. The next time anyone has to deal with partitioning descriptors, I hope they look at this parallel implementation to figure out what the hell is going on.
</blockquote></details>

I don't have strong feelings about this, because that stuff is hard to understand and re-using portions of it wouldn't be very clean, but it's slightly sad that we have 2 very similar implementations. 
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/implicit_partitioning, line 10 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7G4kr9SdObGzj30EM:-Mw7G4kr9SdObGzj30EN:bj5j1ov) ([raw file](https://github.com/cockroachdb/cockroach/blob/40c01c1e9d71a4df691bb9592e841f0e25732138/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/implicit_partitioning#L10)):*
> ```
> 
> exec-sql
> CREATE TABLE db.partition_all_by_list (N1 INT PRIMARY KEY, N2 INT) PARTITION ALL BY LIST (N1) (
> ```

Should you be partitioning on N2 to get implicit partitioning here? Ditto below. 
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions, line 20 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7Gp6r7FK_It4Bhr4T:-Mw7Gp6r7FK_It4Bhr4U:b-5r6suh) ([raw file](https://github.com/cockroachdb/cockroach/blob/40c01c1e9d71a4df691bb9592e841f0e25732138/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions#L20)):*
> ```
> 
> # Partitioning the primary table by range -- we should observe splits along
> # named values within the primary index.
> ```

nit: s/primary table/primary index/g
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions_secondary, line 2 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7MVffB1rXqI2Mfjav:-Mw7MVffB1rXqI2Mfjaw:b3mg8hy) ([raw file](https://github.com/cockroachdb/cockroach/blob/40c01c1e9d71a4df691bb9592e841f0e25732138/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions_secondary#L2)):*
> ```
> # Test a bunch of partitioning variants on secondary indexes. This file mirrors
> # the 'partitions' file. See comments there for more details.
> ```

Consider just bringing them in-line, the readability benefits are worth the duplication :stuck_out_tongue:  
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions_secondary, line 72 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7HPXU-K0WhJYDoE2t:-Mw7HPXU-K0WhJYDoE2u:bb76x64) ([raw file](https://github.com/cockroachdb/cockroach/blob/40c01c1e9d71a4df691bb9592e841f0e25732138/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions_secondary#L72)):*
> ```
> /Table/108/3
> 
> # XXX:
> ```

XXX?
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/datadriven_test.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvv1U4hCaZ7VYwdNoFU:-Mw7Dkes-S_M9LrTzPgh:b-5hqn7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/ccl/spanconfigccl/spanconfigsplitterccl/datadriven_test.go#L35)):*
> Speaking of the SQLTranslator, have we compared its output to that of the Splitter? Are the number of splits/split points the same if we set zone configurations on all indexes/partitions/subpartitions? Thoughts on testing something like this?

Were you meaning to do this part? :) 


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-:-Mw7VUAM5neqCE-YDQO8:b-83330m)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @nvanbenschoten)

___
*[pkg/spanconfig/spanconfig.go, line 300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvv10AQALtuTszoXNoj:-Mw7Niqc0zEp0LSNVbBB:b-ymitd1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfig.go#L300)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Let's change the interface here to return only a count and move the current implementation and testing a layer below it. All usages of this thing are going to be of the form `len(Splitter.Splits())` (they are, right?) so that feels cleaner. 

Also, could you please move the TODO around the alternative implementation to package `spanconfigsplitter` now that we're not going with that approach and it's an implementation detail? 
</blockquote></details>

Moved the TODO. I like this interface, even if the current usage is going to only solicit length. I want to be able to test at the level of the interface in future commits instead of hiding it -- this does not feel unclean. I was actually thinking to adding more assertions/verbose logging at callsites to log the differences in split points, in case it comes handy.
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 66 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-MvuxUdu9xOCnuIicOVw:-Mw7OEOT2kJv2AsFcetg:b-5ly9eu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L66)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I was thinking we might be over counting if we count splits for indexes that are being dropped. Maybe that's okay, considering the drop can fail, and you don't want to run into limits when that happens. 
</blockquote></details>

Will learn more about schema mutations and what needs to change here once wiring it up to the schema changer. I definitely don't understand all corners of what can be dropped, what can be recovered, and what splits to preserve vs. not. Can revisit this things once it's more integrated into running code.
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 133 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvv-gPx4Lgtb_U2TstZ:-Mw7OczaCNICVjTeexbm:bpmt05l) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L133)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I don't have strong feelings about this, because that stuff is hard to understand and re-using portions of it wouldn't be very clean, but it's slightly sad that we have 2 very similar implementations. 
</blockquote></details>

I'm not sad about it, I think it's great to have a more understandable (to me) implementation.
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/implicit_partitioning, line 10 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7G4kr9SdObGzj30EM:-Mw7VKic0XNj4lld3cia:b-807mtf) ([raw file](https://github.com/cockroachdb/cockroach/blob/40c01c1e9d71a4df691bb9592e841f0e25732138/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/implicit_partitioning#L10)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should you be partitioning on N2 to get implicit partitioning here? Ditto below. 
</blockquote></details>

Done, though I don't full understand implicit partitioning. The splits look identical before and after the change, does that make sense to you?
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions_secondary, line 2 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7MVffB1rXqI2Mfjav:-Mw7RCA50RAEvFCyJwaz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/40c01c1e9d71a4df691bb9592e841f0e25732138/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions_secondary#L2)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Consider just bringing them in-line, the readability benefits are worth the duplication :stuck_out_tongue:  
</blockquote></details>

Done.
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions_secondary, line 72 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7HPXU-K0WhJYDoE2t:-Mw7RCc-2lcYStFIKb1X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/40c01c1e9d71a4df691bb9592e841f0e25732138/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions_secondary#L72)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

XXX?
</blockquote></details>

Done.
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/datadriven_test.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvv1U4hCaZ7VYwdNoFU:-Mw7RE0EAzj5Vy2oAniq:bytt3y7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/ccl/spanconfigccl/spanconfigsplitterccl/datadriven_test.go#L35)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

> Speaking of the SQLTranslator, have we compared its output to that of the Splitter? Are the number of splits/split points the same if we set zone configurations on all indexes/partitions/subpartitions? Thoughts on testing something like this?

Were you meaning to do this part? :) 

</blockquote></details>

Oh sorry, forgot to respond. I don't think I want to, it's a bit involved and while I agree it's beneficial and would add good coverage by comparing "test impls", it's work and I'm lazy. I did however add tests to the translator in this PR for more partitioning examples and eye-balled the differences (there aren't any) so I'm happy for now.


<!-- Sent from Reviewable.io -->

irfansharif commented:
How we allocate index IDs changed after #73878, which I just rebased on top of. Had to update some data files is all.

ajwerner reviewed (commented):
Just little stuff from me. I think there might be some minor details to work out but mod that, <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 4 of 15 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-:-Mw7xgoE7LQ0GFbgaLB6:b-ot3tx7)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @irfansharif, and @nvanbenschoten)

___
*[pkg/spanconfig/spanconfig.go, line 256 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7x7YaDeWTLiS-3iuc:-Mw7x7YaDeWTLiS-3iud:bhl9hb8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1346d2905bb1d36f6d1baaae7466f0a79206d675/pkg/spanconfig/spanconfig.go#L256)):*
> ```Go
> //		/Table/108/1/4
> //		/Table/108/1/5
> //		/Table/108/1/6
> ```

this case is interested. `6` and `default` are in the same partition, right? why would we split that? 



_Code quote:_
```Go

//		/Table/108/1/7
```
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 34 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mvyvp_A0VnGb08CzKan:-Mw4oqO001dBPLrxdQov:bjwa7hh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L34)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Done.

```go
  alloc := &tree.DatumAlloc{}
  _, fromKey, err := rowenc.DecodePartitionTuple(
    alloc, s.codec, part.table, part.index, part, from, part.prefixDatums)
  if err != nil {
    return err
  }
```

```
$ go build -gcflags '-m' github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigsplitter 
[...]
../../../../spanconfig/spanconfigsplitter/splitter.go:141:13: &tree.DatumAlloc{} does not escape
```
</blockquote></details>

:tada: 
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 66 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-MvuxUdu9xOCnuIicOVw:-Mw7ucuL5fdjP4MbffLR:bgitogd) ([raw file](https://github.com/cockroachdb/cockroach/blob/0237341d4c649677f3ac77fdd2d65747ef46b8a1/pkg/spanconfig/spanconfigsplitter/splitter.go#L66)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Will learn more about schema mutations and what needs to change here once wiring it up to the schema changer. I definitely don't understand all corners of what can be dropped, what can be recovered, and what splits to preserve vs. not. Can revisit this things once it's more integrated into running code.
</blockquote></details>

I like being conservative here. What you've got is conservative. 
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 110 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7uwFNCx5V8BHI0R0c:-Mw7uwFNCx5V8BHI0R0d:bc44qb0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1346d2905bb1d36f6d1baaae7466f0a79206d675/pkg/spanconfig/spanconfigsplitter/splitter.go#L110)):*
> ```Go
> 	var tableSpans []roachpb.Span
> 	var lastIndexSpan roachpb.Span
> 	if err := catalog.ForEachIndex(table, catalog.IndexOpts{}, func(index catalog.Index) error {
> ```

I don't know that `ForEachIndex` states a contract of being ordered. Is it possible to not rely on that, or to enforce it in the interface and go audit the implementation? Preference in the moment for the former. 
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 157 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7vHRK0odf1JpxAQus:-Mw7vHRK0odf1JpxAQut:bi9fy04) ([raw file](https://github.com/cockroachdb/cockroach/blob/1346d2905bb1d36f6d1baaae7466f0a79206d675/pkg/spanconfig/spanconfigsplitter/splitter.go#L157)):*
> ```Go
> 					log.Fatalf(ctx, "expected partitions %s and %s to be adjacent", prev, cur)
> 				}
> 			}
> ```

Can we just turn this into an `AssertionFailedError` and then consider doing something more drastic under test?



_Code quote:_
```Go
		{ // TODO(irfansharif): Should we only assert under test builds?
			if len(indexPartitionSpans) == 0 {
				log.Fatal(ctx, "expected non-zero index partition spans")
			}
			sort.Sort(roachpb.Spans(indexPartitionSpans))
			if firstKey := indexPartitionSpans[0].Key; !firstKey.Equal(currentIndexSpan.Key) {
				log.Fatalf(ctx, "expected first partition key (%s) to match index start (%s)", firstKey, currentIndexSpan.Key)
			}
			if endKey := indexPartitionSpans[len(indexPartitionSpans)-1].EndKey; !endKey.Equal(currentIndexSpan.EndKey) {
				log.Fatalf(ctx, "expected last partition end key (%s) to match index end (%s)", endKey, currentIndexSpan.EndKey)
			}
			for i := range indexPartitionSpans {
				if i == 0 {
					continue
				}
				if prev, cur := indexPartitionSpans[i-1], indexPartitionSpans[i]; !prev.EndKey.Equal(cur.Key) {
					log.Fatalf(ctx, "expected partitions %s and %s to be adjacent", prev, cur)
				}
			}
```
___
*[pkg/spanconfig/spanconfigtestutils/spanconfigtestcluster/tenant_state.go, line 187 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7w9psExN5Vj136xih:-Mw7w9psExN5Vj136xii:bhwjyc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3317ca86bcaa4e40d519a3ee2cc085ccbf989f81/pkg/spanconfig/spanconfigtestutils/spanconfigtestcluster/tenant_state.go#L187)):*
> ```Go
> 			desc, err = descsCol.GetMutableDatabaseByName(ctx, txn, dbName,
> 				tree.DatabaseLookupFlags{
> 					Required:       true,
> ```

without this flag this can return nil without finding the database, right? 
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/implicit_partitioning, line 13 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7wOkw4oFVaJSoWO53:-Mw7wOkw4oFVaJSoWO54:b7salbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1346d2905bb1d36f6d1baaae7466f0a79206d675/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/implicit_partitioning#L13)):*
> ```
>    PARTITION one VALUES IN (1),
>    PARTITION two VALUES IN (2)
> );
> ```

thoughts on using strings as opposed to integers to make the key boundaries and gaps more interesting? 



_Code quote:_
```
CREATE TABLE db.partition_all_by_list (N1 INT PRIMARY KEY, N2 INT) PARTITION ALL BY LIST (N2) (
   PARTITION one VALUES IN (1),
   PARTITION two VALUES IN (2)
);
```
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions, line 90 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7xQ_VEQhGjeXCdA3f:-Mw7xQ_VEQhGjeXCdA3g:bnohowz) ([raw file](https://github.com/cockroachdb/cockroach/blob/1346d2905bb1d36f6d1baaae7466f0a79206d675/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions#L90)):*
> ```
> CREATE INDEX idx ON db.list_partitions (j);
> ALTER INDEX db.list_partitions@idx PARTITION BY LIST (j) (
>   PARTITION one_and_five    VALUES IN (2, 5),
> ```

all of your values tests seem to have non-contiguous values, Can you put in some cases with contiguous values? Also cases with contiguous values mixed with default. 


<!-- Sent from Reviewable.io -->

irfansharif reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75803)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @nvanbenschoten)

___
*[pkg/spanconfig/spanconfig.go, line 256 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7x7YaDeWTLiS-3iuc:-Mw8NQVm6CUo8oc4_5dd:b4u3n0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/1346d2905bb1d36f6d1baaae7466f0a79206d675/pkg/spanconfig/spanconfig.go#L256)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this case is interested. `6` and `default` are in the same partition, right? why would we split that? 
</blockquote></details>

It's how we also do sub-zone spans apparently. And also how we translate constituent zone configs. Running this same test through the sql translator (checked it in as well):
```
exec-sql
CREATE TABLE db.partition_by_list(i INT PRIMARY KEY, j INT) PARTITION BY LIST (i) (
  PARTITION one_and_five    VALUES IN (1, 5),
  PARTITION four_and_three  VALUES IN (4, 3),
  PARTITION everything_else VALUES IN (6, default)
);
ALTER TABLE db.partition_by_list CONFIGURE ZONE USING gc.ttlseconds = 1;
ALTER PARTITION one_and_five OF TABLE db.partition_by_list CONFIGURE ZONE USING gc.ttlseconds = 2;
ALTER PARTITION four_and_three OF TABLE db.partition_by_list CONFIGURE ZONE USING gc.ttlseconds = 3;
ALTER PARTITION everything_else OF TABLE db.partition_by_list CONFIGURE ZONE USING gc.ttlseconds = 4;
----

translate database=db table=partition_by_list
----
/Table/109{-/1}                            ttl_seconds=1
/Table/109/1{-/1}                          ttl_seconds=4
/Table/109/1/{1-2}                         ttl_seconds=2
/Table/109/1/{2-3}                         ttl_seconds=4
/Table/109/1/{3-4}                         ttl_seconds=3
/Table/109/1/{4-5}                         ttl_seconds=3
/Table/109/1/{5-6}                         ttl_seconds=2
/Table/109/1/{6-7}                         ttl_seconds=4
/Table/109/{1/7-2}                         ttl_seconds=4
/Table/109/{2-PrefixEnd}                   ttl_seconds=1
```

Note the two distinct entries for the following:
```
/Table/109/1/{6-7}                         ttl_seconds=4
/Table/109/{1/7-2}                         ttl_seconds=4
```

In all our uses of sub-partitions, we do the simple thing and partition on explicitly listed values even if they're co-listed with DEFAULTs. The splitter is not different (good or bad thing, I don't know -- but same thing for accounting reasons).
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 110 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7uwFNCx5V8BHI0R0c:-Mw8O8eYCjvEKvbm_rTJ:b4prj74) ([raw file](https://github.com/cockroachdb/cockroach/blob/1346d2905bb1d36f6d1baaae7466f0a79206d675/pkg/spanconfig/spanconfigsplitter/splitter.go#L110)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I don't know that `ForEachIndex` states a contract of being ordered. Is it possible to not rely on that, or to enforce it in the interface and go audit the implementation? Preference in the moment for the former. 
</blockquote></details>

Done -- I'm assuming you meant key layout ordering of indexes, right? It does go through them in ordinal order, which I realize now isn't the same thing. Great catch!
___
*[pkg/spanconfig/spanconfigsplitter/splitter.go, line 157 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7vHRK0odf1JpxAQus:-Mw8PMwiDe9J0exhx04H:b-tefcy4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1346d2905bb1d36f6d1baaae7466f0a79206d675/pkg/spanconfig/spanconfigsplitter/splitter.go#L157)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Can we just turn this into an `AssertionFailedError` and then consider doing something more drastic under test?
</blockquote></details>

Done. Under test it'll fail informatively -- were you thinking something different?
___
*[pkg/spanconfig/spanconfigtestutils/spanconfigtestcluster/tenant_state.go, line 187 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7w9psExN5Vj136xih:-Mw8Q-ye7X418VK5DuhQ:b-d7ybkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/3317ca86bcaa4e40d519a3ee2cc085ccbf989f81/pkg/spanconfig/spanconfigtestutils/spanconfigtestcluster/tenant_state.go#L187)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

without this flag this can return nil without finding the database, right? 
</blockquote></details>

Yea, but we can use this flag (Done). I think this change was accidental debugging detritus after seeing broken name resolution for dropped descriptors.
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/implicit_partitioning, line 13 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7wOkw4oFVaJSoWO53:-Mw8RW7lDU6cRBueXkD_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1346d2905bb1d36f6d1baaae7466f0a79206d675/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/implicit_partitioning#L13)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

thoughts on using strings as opposed to integers to make the key boundaries and gaps more interesting? 
</blockquote></details>

Done.
___
*[pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions, line 90 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/75803#-Mw7xQ_VEQhGjeXCdA3f:-Mw8SIkr3D8-9wjbl8as:b-jx2u38) ([raw file](https://github.com/cockroachdb/cockroach/blob/1346d2905bb1d36f6d1baaae7466f0a79206d675/pkg/ccl/spanconfigccl/spanconfigsplitterccl/testdata/partitions#L90)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

all of your values tests seem to have non-contiguous values, Can you put in some cases with contiguous values? Also cases with contiguous values mixed with default. 
</blockquote></details>

There's a ` PARTITION four_and_three  VALUES IN (4, 3),` in here which is contiguous, but I assume you meant partitioning _only_ by contiguous values? Added.


<!-- Sent from Reviewable.io -->

irfansharif commented:
bors r+
